### PR TITLE
Improved Index Map Features, Multiply function

### PR DESCRIPTION
**Pull request by JoeyT1994**
This PR adds some significant changes to the `IndexMap` and `IndexNetworkMap` interfaces as well as changes to the `multiply()` function.

- `rem_index(imap::ComplexIndexMap, ind::Index)` and `rem_index(imap::RealIndexMap, ind::Index)` is now defined to remove indices from the dictionaries inside `IndexMaps`.
- `dimensions(imap::AbstractIndexMap)` is now defined and returns a vector of all the dimensions of the indices in the map. An `imap` is not required to have a strictly sequential list of dimensions and could just have indices `ind` with `dimension(imap, ind) = 2` and nothing else.
- `reduced_indexmap(imap::AbstractIndexMap, dims::Vector{<:Int})` is defined and returns a new index map with just the indices in the given dimensions
- `merge(imap1::RealIndexMap, imap2::RealIndexMap)` and `merge(imap1::ComplexIndexMap, imap2::ComplexIndexMap)`  is defined and creates a new index map based on the merging of all the dictionaries inside the arguments.
- `reduced_indsnetworkmap(inm::IndsNetworkMap, dims::Vector{<:Int})` is defined and returns a new `indexnetworkmap` based on just the vertices/ indices which are in the corresponding dimension. The resulting graph in `indsnetwork(inm)` will thus have changed and is that obtained upon removing all vertices which don't have any external indices in `dims` 
- `multiply(f::ITensorNetworkFunction, g::ITensorNetworkFunction)` has been re-written so there is no assumption the underlying graphs are the same. Only external indices which are common between `f` and `g` are treated as identical and a direct product is taken over everything else. This allows one to take a very general "multiplication" like `h(x,y) = f(x) * g(y)` where `f` and `g` live in completely different dimensions  and on different graphs to create a new tensor network which is the merging of the two graphs
- `partial_integrate(fits::ITensorNetworkFunction, dims::Vector{<:Int}; merge_vertices = true)` is now defined and integrates only over the specified dimensions. The flag `merge_vertices` specifies whether to contract away the tensors that were integrated over to create a new `ITensorNetworkFunction` which lives on a reduced graph.

@ryanlevy The Fredholm example has been updated to reflect the utility of these changes. Now `ψ` is always reduced to a one-dimensional function following the partial integration. And you can just call `evaluate(ψ, x::Number)` to evaluate it and it doesn't matter whether you technically ended up in the `y` or `x` dimension  at the end because `ψ` only has indices that live in one dimension. Because the Kernel is separable the bond dimension of `ψ` is always bounded as `chi <= 3` (the bond dimension of the kernel `g` + the bond dimension of the additive function `c`) in this example without any calls to truncate which I think is really neat and exposes that the code is working well. 

Note that I had to overload the `add(itn1::ITensorNetwork, itn2::ITensorNetwork)` function from `ITensorNetworks` here because the `merge_vertices` operation reverses some edges in the network and `add` currently asserts too strictly that the edges have to be ordered identically to work. This is a bug that needs to be fixed in `ITensorNetworks` but I put a fix here for now.

### CODE REVIEW

File: src/integration.jl [hunk 1]
```diff
@@ -25,3 +25,23 @@ function integrate(
   ∑ = ITensorNetwork(eltype(first(fitn)), v -> [0.5, 0.5], s)
   return inner(∑, operator, itensornetwork(fitn); alg, kwargs...)
 end
+
+function partial_integrate(
+  fitn::ITensorNetworkFunction, dims::Vector{Int}; merge_vertices=true
+)
```
Comment (ryanlevy) [inline src/integration.jl:31]: Could we add a docstring?

**Review by ryanlevy — CHANGES_REQUESTED** [review]
Thanks Joey, this largely looks good! One minor request, could you please add a test for `partial_integrate`?

File: src/integration.jl [hunk 2]
```diff
@@ -25,3 +25,23 @@ function integrate(
...
+  s = indsnetworkmap(fitn)
+  new_imap = copy(indexmap(s))
+  fitn = copy(itensornetwork(fitn))
+  for v in dimension_vertices(s, dims)
+    sinds_dim = filter(i -> dimension(s, i) ∈ dims, s[v])
+    for sind in sinds_dim
+      fitn[v] *= ITensor(eltype(fitn[v]), 0.5, sind)
+      new_imap = rem_index(new_imap, sind)
+    end
```
[the remark this answers is not in the document]
Comment (JoeyT1994) [reply src/integration.jl]: Yeah it does. I changed it to support any base.

**Comment by JoeyT1994** [comment]
Test added.

**Review by ryanlevy — APPROVED** [review]
LGTM now, thanks!

**Pull request merged**